File: src/modules/AlgoDatCourse/implemented/java/MinHeap.java
import java.util.Arrays;

/**
 * Eingabe:    Ein unsortiertes Array A von Zahlen.
 * Ausgabe:    Ein Array A, das als Min-Heap organisiert ist.
 */



public class MinHeap {

    private int[] A;
    private int heapSize;

    MinHeap(int[] A) {
        this.A = A;
        this.heapSize = A.length;
        buildMinHeap();
    }

    /** HAUPTMETHODEN */

    public void buildMinHeap() {
        heapSize = A.length;

        for (int i = heapSize / 2 - 1; i >= 0; i--) {
            minHeapify(i);
        }
    }

    public void minHeapify(int i) {
        int l = left(i);
        int r = right(i);

        int smallest;

        if (l < heapSize && A[l] < A[i]) {
            smallest = l;
        } else {
            smallest = i;
        }

        if (r < heapSize && A[r] < A[smallest]) {
            smallest = r;
        }

        if (smallest != i) {
            int temp = A[i];
            A[i] = A[smallest];
            A[smallest] = temp;

            minHeapify(smallest);
        }
    }

    public void minHeapInsert(int key) {
        heapSize = A.length + 1;

        A = Arrays.copyOf(A, heapSize);

        A[heapSize - 1] = Integer.MAX_VALUE;

        minHeapDecreaseKey(heapSize - 1, key);
    }

    public int extractMin() {
        if (heapSize == 0) {
            throw new IllegalStateException("Heap is empty");
        }

        int min = A[0];

        A[0] = A[heapSize - 1];

        heapSize--;

        A = Arrays.copyOf(A, heapSize);

        if (heapSize > 0) {
            minHeapify(0);
        }

        return min;
    }

    public void minHeapDecreaseKey(int i, int key) {

        if (key > A[i]) {
            throw new IllegalArgumentException(
                "Error: new key is larger than current key"
            );
        }

        A[i] = key;

        while (i > 0 && A[parent(i)] > A[i]) {

            int temp = A[i];
            A[i] = A[parent(i)];
            A[parent(i)] = temp;

            i = parent(i);
        }
    }

    /** HILFSFUNKTIONEN IM BAUM */

    private int left(int i) {
        return 2 * i + 1;
    }

    private int right(int i) {
        return 2 * i + 2;
    }

    private int parent(int i) {
        return (i - 1) / 2;
    }

    /** GETTER METHODEN */

    public int[] getHeap() {
        return A;
    }

    public int[] getSortedArray() {

        int[] backupA = Arrays.copyOf(A, heapSize);
        int backupHeapSize = heapSize;

        int[] sortedArray = new int[heapSize];

        for (int i = 0; i < sortedArray.length; i++) {
            sortedArray[i] = extractMin();
        }

        A = backupA;
        heapSize = backupHeapSize;

        return sortedArray;
    }

    public int getMinimum() {

        if (heapSize == 0) {
            throw new IllegalStateException("Heap is empty");
        }

        return A[0];
    }

    public int getMaximum() {

        if (heapSize == 0) {
            throw new IllegalStateException("Heap is empty");
        }

        int max = A[heapSize / 2];

        for (int i = heapSize / 2; i < heapSize; i++) {

            if (A[i] > max) {
                max = A[i];
            }
        }

        return max;
    }

    public static void main(String[] args) {

        MinHeap minHeap = new MinHeap(new int[]{5, 2, 1, 3, 7, 8});

        System.out.println(Arrays.toString(minHeap.getHeap()));

        System.out.println(Arrays.toString(minHeap.getSortedArray()));

        System.out.println(minHeap.getMinimum());

        System.out.println(minHeap.getMaximum());
    }
}

File: src/modules/AlgoDatCourse/implemented/python/MinHeap.py
class MinHeap:

    def __init__(self, A):

        self.A = A
        self.heapSize = len(A)

        self.buildMinHeap()

    """ HAUPTMETHODEN """

    def buildMinHeap(self):

        self.heapSize = len(self.A)

        for i in range(self.heapSize // 2 - 1, -1, -1):
            self.minHeapify(i)

    def minHeapify(self, i):

        l = self.left(i)
        r = self.right(i)

        if l < self.heapSize and self.A[l] < self.A[i]:
            smallest = l
        else:
            smallest = i

        if r < self.heapSize and self.A[r] < self.A[smallest]:
            smallest = r

        if smallest != i:

            temp = self.A[i]
            self.A[i] = self.A[smallest]
            self.A[smallest] = temp

            self.minHeapify(smallest)

    def minHeapInsert(self, key):

        self.heapSize = len(self.A) + 1

        self.A.append(float("inf"))

        self.minHeapDecreaseKey(self.heapSize - 1, key)

    def extractMin(self):

        if self.heapSize == 0:
            raise Exception("Heap is empty")

        minimum = self.A[0]

        self.A[0] = self.A[self.heapSize - 1]

        self.heapSize -= 1

        self.A.pop()

        if self.heapSize > 0:
            self.minHeapify(0)

        return minimum

    def minHeapDecreaseKey(self, i, key):

        if key > self.A[i]:
            raise Exception(
                "Error: new key is larger than current key"
            )

        self.A[i] = key

        while i > 0 and self.A[self.parent(i)] > self.A[i]:

            temp = self.A[i]
            self.A[i] = self.A[self.parent(i)]
            self.A[self.parent(i)] = temp

            i = self.parent(i)

    """ HILFSFUNKTIONEN IM BAUM """

    def left(self, i):
        return 2 * i + 1

    def right(self, i):
        return 2 * i + 2

    def parent(self, i):
        return (i - 1) // 2

    """ GETTER METHODEN """

    def getHeap(self):
        return self.A

    def getSortedArray(self):

        backupA = self.A.copy()
        backupHeapSize = self.heapSize

        sortedArray = [0] * self.heapSize

        for i in range(len(sortedArray)):
            sortedArray[i] = self.extractMin()

        self.A = backupA
        self.heapSize = backupHeapSize

        return sortedArray

    def getMinimum(self):

        if self.heapSize == 0:
            raise Exception("Heap is empty")

        return self.A[0]

    def getMaximum(self):

        if self.heapSize == 0:
            raise Exception("Heap is empty")

        maximum = self.A[self.heapSize // 2]

        for i in range(self.heapSize // 2, self.heapSize):

            if self.A[i] > maximum:
                maximum = self.A[i]

        return maximum


if __name__ == "__main__":

    minHeap = MinHeap([5, 2, 1, 3, 7, 8])
    print(minHeap.getHeap())
    print(minHeap.getSortedArray())
    print(minHeap.getMinimum())
    print(minHeap.getMaximum())